service/controller/resource/nodes/clients.go

```diff
@@ -0,0 +1,57 @@
+package nodes
```

Comment:
Sorry, I know I'm a pain in the ass, but I'm not fully convinced into this approach.

It's harder for me to understand the code now with the "inherited" fields. I think it will be harder to test and modify, since changes on the "parent" will affect on both resources now. Specially when it seems master will differ from workers in the near future.

I was thinking about using composition creating objects that allow us to reuse functionality. For example, if we are going to move away of controllercontext in the future, I think it would be a good idea to have a factory that returns azure clients. The logic would live on the factory. Resources just depend on the factory to fetch their clients.
I'd propose something similar for the `GetEncrypterObject` function.

For functions like `AllInstances` or `AnyOutOfDate` we could place them as independent functions in the `vmss` package? (`pkg/helpers/vmss`)

Does it make sense?

This way we could test each of those pieces in isolation and re-use them wherever needed.

Reply:
> with the "inherited" fields.
> I was thinking about using composition creating objects that allow us to reuse functionality.

Technically speaking, this **is** composition :smiley:  i.e. embedding one type into another, there is no "inheritance" :) It's just golang's syntactic sugar that let's you use fields and functions from the `nodes.Resource` type as if they were in `masters.Resource` / `instance.Resource`.

Another thing that led me to this approach is the fact that `masters.Resource`/`masters.Config` and `instance.Resource`/`instance.Config` are 100% identical structures.

> I think it will be harder to test and modify, since changes on the "parent" will affect on both resources now. Specially when it seems master will differ from workers in the near future.

The idea is that you do not have to think about masters and workers here at all, it's just some set of nodes/machines. Basically separating generic VMs/VMSS/instances/nodes stuff into a single shared package, so that we can reuse it and not copy-paste same code around.

I even had a bit "conservative" approach here, for example:
- `masters/deployment.go` and `instance/deployment.go` are basically identical, it's practically a parametrized VMSS deployment (the only difference is resource naming)
- some reconciliation functions are still the same/identical, for example `DeployementUninitialized`, `DeployementInitialized`, `ProvisioningSuccessful`, `DeploymentCompleted` (and probably few more) transitions are basically copy-paste with only difference around Azure resource naming.

:point_up_2: these could also be deduplicated easily, since they are about reconciliation of a VMSS deployment, nothing masters-/workers-specific there, it's just deployment of compute things.

> For example, if we are going to move away of controllercontext in the future, I think it would be a good idea to have a factory that returns azure clients. The logic would live on the factory. Resources just depend on the factory to fetch their clients.

Some stuff now ended up here just because it was duplicated in `masters` and `instance` packages.

Yeah, clients can definitely be moved outside of controllers/resources into some factory. Will do that and update this PR.

> I'd propose something similar for the GetEncrypterObject function.

`GetEncrypterObject` is not really "static"/global, it's using `Logger` and `k8sClient` fields of the resource, so that would require a bit more refactoring if we were to move it outside of `nodes` package.

> For functions like AllInstances or AnyOutOfDate we could place them as independent functions in the vmss package? (pkg/helpers/vmss)

`nodes` package is supposed to be a kind of "helpers" package, but resource/reconciliation aware. Other option for name I had in mind was `machines`, maybe that would be better since `masters` and `instance` packages are about both Kubernetes nodes and Azure VMSS/instances, either `nodes` or `vmss` is maybe kind of limiting.
I'll take a look into `pkg/helpers/vmss` package.

Generally speaking, with splitting `masters` and `instance` into separate resources, we do have separate deployments now, which is much better, but good part of reconciliation is still largely the same, and once we remove Flatcar-related workarounds, the difference will be even less.

IMHO, when it comes to nodes/VMSS reconciliation, ideally we would have these 3 layers of packages, where the each layer is using the one above/before it:
1. nodes/machines services, factories, utilities, basically a set of functionalities for working with Kubernetes nodes and Azure VMSS (e.g. you should not write multiple times the same code to check if the nodes are creating, out of date, etc.)
2. `Azure VMSS` <-> `Kubernetes nodes` reconciliation: e.g. you should not write multiple times the same code to check if a VMSS deployment has changed, or to create/update a deployment, etc.
3. `masters` and `workers`/`instance` packages, with `masters` and `workers` specific reconciliation.

This would kind of make us think what we want to change. Is it about some Azure client thing, some node state check -> layer 1 above. Reconciling some new flag/property for Azure VMSS -> layer 2. Some etcd related change -> `masters` in layer 3, etc.

On the other hand, if we move away from VMSS to VMs for masters, than that is a different story, but it's questionable if and when will that happen :)

Reply:
In this 3-layer separation, layer 1 would be about talking to Kubernetes and Azure. While layers 2 and 3 would be about deciding when to talk and who to talk to.

Also if we are not sure if some functionality/behaviour is shared between masters and workers, or only specific to masters or workers only, then I would say we need to refine that task to make that clear.

Reply:
> While layers 2 and 3 would be about deciding when to talk and who to talk to.

For example, when updating a VMSS deployment, we should not care about some ARM template or deployment specifics, ideally we would have some simple code with some simple logic, focus on reconciliation logic.

Let's take `deploymentInitializedTransition` for example, it comes down to:
```golang
if deployment.IsSucceeded {
    // all done, continue
    return ProvisioningSuccessful
} else if deployment.IsError {
    // error, start from the scratch
    return DeploymentUninitialized
} else {
    // still working
    return currentState
}
```

Switch would maybe make it even more straightforward :)

And it should be basically as simple as that, and not repeating that code every time you want to check a deployment and decide what to do next.

`deploymentUninitializedTransition` is not much different (some other as well), in terms of having a simple reconciliation logic (_layers 2 and 3_), with non-reconciliation logistics (_layer 1_) being out if your hair.

Reply:
I like the direction where this is going. Few thoughts though:
- Can we limit changes to some manageable size here and settle with state that we can take further in next steps, without fully refactoring everything now?
- What do you think if we actually tried to implement the node pools resources in TDD fashion  in the sense that we try to create some kind of integration and unit tests for the whole implementation while actually doing it? I figure that would help us in the future a lot and might help to solve some concerns here.
- The client factory idea is something we need to drive. That solves the problems I had with metrics collection, but it also makes it easier to implement caching for it as well.
- I wouldn't place too much weight on masters management implementation right now because I think no matter what route we take to HA masters, it will be blown up mostly anyway and won't stay how it is right now.

Reply:
> Can we limit changes to some manageable size here and settle with state that we can take further in next steps, without fully refactoring everything now?

Yes! This was more about the code deduplication, less about some bigger refactoring.

> What do you think if we actually tried to implement the node pools resources in TDD fashion in the sense that we try to create some kind of integration and unit tests for the whole implementation while actually doing it? I figure that would help us in the future a lot and might help to solve some concerns here.

Interesting idea, let's talk about it.

> The client factory idea is something we need to drive. That solves the problems I had with metrics collection, but it also makes it easier to implement caching for it as well.

Agreed. Now we are getting the clients from `controllercontext`, so I'm looking into that, to see if we can decouple Azure clients from `controllercontext`.